lib/module-utils.js

```diff
@@ -0,0 +1,171 @@
+const eslint = require('eslint')
+const fs = require('fs')
+const minimatch = require('minimatch')
+const moduleVisitor = require('eslint-module-utils/moduleVisitor')
+const parse = require('eslint-module-utils/parse').default
+const resolve = require('eslint-module-utils/resolve').default
+
+exports.detectUsedModuleExports = function(files) {
+  return Promise.all(files.map(exports.load)).then(results => {
+    const imported = new Map()
+
+    for (const {path, ignore, exports} of results) {
+      if (ignore) {
+        continue
+      }
+
+      const usedExports = new Map()
+      imported.set(path, usedExports)
+      for (const identifier of exports) {
+        usedExports.set(identifier, false)
+      }
+    }
+
+    for (const {imports} of results) {
+      for (const [importedPath, importedIdentifiers] of imports) {
+        if (imported.has(importedPath)) {
+          for (const importedIdentifier of importedIdentifiers) {
+            if (importedIdentifier === '*') {
+              imported.set(importedPath, true)
+            } else {
+              imported.get(importedPath).set(importedIdentifier, true)
+            }
+          }
+        }
+      }
+    }
+
+    const unusedImports = new Map()
+
+    for (const [path, exports] of imported) {
+      if (exports === true) {
+        continue
+      } else if (exports.size === 0) {
+        unusedImports.set(path, true)
+      } else {
+        const unusedIdentifiers = new Set()
+        for (const [identifier, used] of exports) {
+          if (!used) unusedIdentifiers.add(identifier)
+        }
+        if (unusedIdentifiers.size !== 0) {
+          unusedImports.set(path, unusedIdentifiers)
+        }
+      }
+    }
+
+    return unusedImports
+  })
+}
+
+exports.load = function(path) {
+  return readFile(path, {encoding: 'utf8'}).then(content => exports.parse(path, content))
+}
+
+exports.parse = function(path, content) {
+  const result = {
+    path,
+    entry: false,
+    imports: new Map(),
+    exports: new Set()
+  }
+
+  const cli = new eslint.CLIEngine()
+  const options = cli.getConfigForFile(path)
```

Comment:
How expensive is doing `getConfigForFile` for every individual file? Is this what eslint does normally as well?

Reply:
I think this needs to be revised. It is expensive on our larger internal codebases.